src/shared/libfido2-util.c

```diff
@@ -521,6 +531,11 @@ int fido2_generate_hmac_hash(
                                        "Requested to lock with PIN, but FIDO2 device %s does not support it.",
                                        device);
 
+        if (!has_up && lock_with_presence)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Requested to lock with user presence, but FIDO2 device %s does not support it.",
+                                       device);
+
```

Comment:
we should set sym_fido_assert_set_up() based on lock_with_presence I guess, further down, so that the assert config matches exactly what is encoded in the JSON data

Reply:
it is already doing that:

```
        if (r == FIDO_ERR_UP_REQUIRED || (lock_with_presence && has_up)) {

                if (!has_up)
                        log_warning("Weird, device asked for User Presence check, but does not advertise it as feature. Ignoring.");

                r = sym_fido_assert_set_up(a, FIDO_OPT_TRUE);
```

Or did you mean something else?

Reply:
What I am saying is that we should tweak up/uv/clientPin anymore after we tried once. we should store precisely what up/uv/clientPin was set to when we enrolled it and then rebuild the assertion from that, and try exactly once with that data, and not try any other config anymore if that fails

Reply:
ok, got it now, thanks